### PR TITLE
Add support for SiLabs EFR32MG21 SoC

### LINKED ISSUE #16968: silabs/gecko/emlib/src/em_gpio.c:111:35: warning: ?: using integer constants in boolean context [-Werror=int-in-bool-context]
**Describe the bug**
[ccache used to cache pre-processor output as a speed optimization](https://github.com/ccache/ccache/issues/116). But contrary to a popular myth, C compilers don't handle already pre-processed code the same. Feeding the compiler with already pre-processed code surfaces one hidden warning in `silabs/gecko/emlib/src/em_gpio.c`. This used to be a sanitycheck showstopper because sanitycheck adds `-Werror`.
```
modules/hal/silabs/gecko/emlib/src/em_gpio.c: In function 'GPIO_DriveModeSet':
modules/hal/silabs/gecko/emlib/src/em_gpio.c:111:35: warning: ?: using integer constants in boolean context [-Werror=int-in-bool-context]
   EFM_ASSERT(GPIO_PORT_VALID(port) && GPIO_DRIVEMODE_VALID(mode));
           ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

Looking at the pre-processed output, the warning is valid (the name GPIO_PORT_VALID is misleading) and easy to fix with some `(bool)` casts. It looks like the warning gets "optimized away" when not using ccache.

**Workarounds**
- Disable ccache - and everything becomes slower
- `rm -rf ~/.ccache; export CCACHE_CPP2=anything_except_yes` and everything becomes slower but just once.

Second workaround and possible explanation found at http://petereisentraut.blogspot.com/2011/09/ccache-and-clang-part-2.html 

**To Reproduce** on any Unix system with ccache
```
mv ~/.ccache ~/ccache.saved
unset CCACHE_CPP2
export CCACHE_NOCPP2=anything
ccache --print-config | grep cpp # (environment) run_second_cpp = false
cd ${ZEPHYR_BASE}/samples/synchronization
rm -rf build/
cmake -B build -S . -DBOARD=efm32hg_slstk3400a -DEXTRA_CFLAGS=-Werror
make -C build VERBOSE=1
```

**Expected behavior**
Compiles.

**Impact**
sanitycheck showstopper

**Environment (please complete the following information):**
 - OS: Ubuntu 16.04, ccache version 3.2.4
 - Toolchain  Zephyr SDK 0.10.0
- https://github.com/zephyrproject-rtos/hal_silabs 9151e614c239

**Additional context**
I verified that ccache doesn't change the .obj file, it only adds the warning.

### PR DESCRIPTION
**Pull request by LeGrosFaisan**
Adds support for Silicon Labs EFR32MG21 (Mighty Gecko Series 2) SoC and BRD4180A (a.k.a. SLWRB4180A) board.

There is a PR for the HAL at [#7](https://github.com/zephyrproject-rtos/hal_silabs/pull/7)

### CODE REVIEW

File: soc/arm/silabs_exx32/Kconfig [hunk 1]
```diff
@@ -81,6 +81,11 @@ config SOC_GECKO_RTCC
 	help
 	  Set if the Real Time Counter and Calendar (RTCC) HAL module is used.
 
+config SOC_GECKO_SE
+	bool
+	help
+	  Set if the Secure Element (SE) HAL module is used.
```
Comment (ioannisg) [inline soc/arm/silabs_exx32/Kconfig:87]: @l-alfred is this related to TrustZone-M ?

Comment (l-alfred) [reply soc/arm/silabs_exx32/Kconfig:87]: @ioannisg it is related to all security features available on the Gecko Series 2 SoC. It comprises a TRNG, hardware cryptographic acceleration elements (e.g., AES, SHA, ECC), the ARM TrustZone, a locking mechanism of the debug functionalities and some secure boot features. It works by sending commands to the SE mailbox (i.e., 64 word FIFO).

For example, It is used by the Silicon Labs mbedTLS implementation.

Comment (ioannisg) [reply soc/arm/silabs_exx32/Kconfig:87]: Is the option used in Zephyr, though?

Comment (ioannisg) [reply soc/arm/silabs_exx32/Kconfig:87]: Does this SoC implements the Security Extension, btw?

Comment (l-alfred) [reply soc/arm/silabs_exx32/Kconfig:87]: > Is the option used in Zephyr, though?

I have an implementation of the entropy driver using the SE option. However, Silicon Labs recommends to use it through their mbedTLS implementation instead. That is why I did not push the driver, as I am not sure what is the best option. In any case, this will be needed for security-related topics why using the SoC, let it be the third-party mbedTLS implementation or a simple entropy driver.

Comment (l-alfred) [reply soc/arm/silabs_exx32/Kconfig:87]: > Does this SoC implements the Security Extension, btw?

I am not sure I follow, what do you mean by Security Extension? Are you referring to the secure-privileged, non-secure-privileged, secure-unprivileged and non-secure-unprivileged states?

Comment (ioannisg) [reply soc/arm/silabs_exx32/Kconfig:87]: I am asking if this SoC implements TrustZone-M (ARM Security Extension for the M profile). That's essential.

- If it does not, then, I don't see any issues.
- It if does, then there should be a selection of a Kconfig that indicates the SAU/IDAU presence. You should also be selecting ARM_TRUSTZONE_M. In that case, is it possible to run Zephyr as Secure firmware on this SoC?
- If TrustZone -M is present, but Zephyr is only supposed to run together with a secure FW counterpart, as a Non-Secure firmware, then, you should be enabling TRUSTED_EXECUTION_NONSECURE by default, and document this in the .rst.

Comment (l-alfred) [reply soc/arm/silabs_exx32/Kconfig:87]: > I am asking if this SoC implements TrustZone-M (ARM Security Extension for the M profile). That's essential.
> 
>     * If it does not, then, I don't see any issues.
> 
>     * It if does, then there should be a selection of a Kconfig that indicates the SAU/IDAU presence. You should also be selecting ARM_TRUSTZONE_M. In that case, is it possible to run Zephyr as Secure firmware on this SoC?
> 
>     * If TrustZone -M is present, but Zephyr is only supposed to run together with a secure FW counterpart, as a Non-Secure firmware, then, you should be enabling TRUSTED_EXECUTION_NONSECURE by default, and document this in the .rst.

I see, although the SoC as the possibility to use the ARM TrustZone-M, this PR does not implement it.

Comment (ioannisg) [reply soc/arm/silabs_exx32/Kconfig:87]: That's OK. Then I suppose Zephyr will run, by default in Secure mode and will not sue any TrustZone-M features. Thanks for clarifying.

Comment (l-alfred) [reply soc/arm/silabs_exx32/Kconfig:87]: Indeed, by default it seems to run in Secure mode. I can try to have a look and implement it, but as I'm not an expert, it might take a while. I think it will come with another PR.

File: soc/arm/silabs_exx32/efr32mg21/soc.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Board configuration macros for the EFR32MG21 SoC
+ *
+ */
+
+#ifndef _SOC__H_
```
Comment (parthitce) [inline soc/arm/silabs_exx32/efr32mg21/soc.h:13]: `ZEPHYR_SOC_ARM_SILABS_EXX32_EFR32MG21_SOC_H_` instead?

File: CODEOWNERS [hunk 3]
```diff
@@ -273,6 +273,7 @@
 /dts/arm/silabs/efr32bg13p*               @mnkp
 /dts/arm/silabs/efm32jg12b*               @chrta
 /dts/arm/silabs/efm32pg12b*               @chrta
+/dts/arm/silabs/efr32mg21*                @l-alfred
```
Comment (parthitce) [inline CODEOWNERS:276]: Also `boards/arm/efr32mg21_slwstk6006a/*` & `soc/arm/silabs_exx32/efr32mg21/*` ?

File: soc/arm/silabs_exx32/efr32mg21/soc.h (hunk 2 above)
Comment (l-alfred) [reply soc/arm/silabs_exx32/efr32mg21/soc.h]: @parthitce Ok done, thanks for the suggestion

**Review by chrta — CHANGES_REQUESTED** [review]
Thank you for your contribution. Overall this looks good. I have not had time to review in more detail.

File: boards/arm/efr32mg21_slwstk6006a/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (l-alfred) [reply boards/arm/efr32mg21_slwstk6006a/CMakeLists.txt]: I've modified the board structure and merged it with the efr32_radio folder, as you mentioned.

**Review by chrta — APPROVED** [review]
LGTM. I performed the following tests and it behaves as in the master branch:
```
sanitycheck -p efm32pg_stk3402a --device-testing --device-serial /dev/ttyACM1
sanitycheck -p efr32_radio_brd4250b --device-testing --device-serial /dev/ttyACM1
```

Before merging this, https://github.com/zephyrproject-rtos/hal_silabs/pull/7 must be merged and west.yml updated.
Also the pr https://github.com/zephyrproject-rtos/zephyr/pull/21292 should be merged first.

File: CODEOWNERS [hunk 5]
```diff
@@ -42,6 +42,7 @@
 /soc/arm/nuvoton/                         @ssekar15
 /soc/arm/qemu_cortex_a53/                 @carlocaione
 /soc/arm/st_stm32/                        @erwango
+/soc/arm/silabs_exx32/efr32mg21/          @l-alfred
```
Comment (MaureenHelm) [inline CODEOWNERS:45]: Please move this up a line so it's not stuck between two st_stm32 entries

File: boards/arm/efr32_radio/doc/index.rst [hunk 6]
```diff
@@ -17,6 +18,7 @@ Support for EFR32 Radio boards is provided by one of the starter kits
 - `SLWSTK6020B Bluetooth SoC Starter Kit`_
 - `SLWSTK6000B Mighty Gecko Wireless Starter Kit`_
 - `SLWSTK6061B Proprietary Wireless Starter Kit`_
+- `SLWSTK6006A Migthy Gecko Wireless Starter Kit`_
```
Comment (MaureenHelm) [inline boards/arm/efr32_radio/doc/index.rst:21]: typo? s/Migthy/Mighty/

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> Before merging this, [zephyrproject-rtos/hal_silabs#7](https://github.com/zephyrproject-rtos/hal_silabs/pull/7) must be merged

This is done now

File: CODEOWNERS (hunk 5 above)
Comment (l-alfred) [reply CODEOWNERS:44]: This is done now, thanks for your input @MaureenHelm

File: boards/arm/efr32_radio/doc/index.rst (hunk 6 above)
Comment (l-alfred) [reply boards/arm/efr32_radio/doc/index.rst]: Ah yes, well spotted! Thanks.

**Review by MaureenHelm — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
